Use rpmdbCookie from librpm, remove hawkey.Sack._rpmdb_version

jrohel opened the pull request:
PR https://github.com/rpm-software-management/dnf/pull/1811 must be merged first

`dnf_sack_get_rpmdb_version` function that computed the hash of installed packages was removed. `rpmdbCookie` function from librpm is used in context part of libdnf instead.

The private Python function `hawkey.Sack._rpmdb_version()` was removed too. It was used by DNF. The new version of DNF uses `rpm.TransactionSet.dbCookie()` function from librpm, which wraps the `rpmdbCookie` function. So the same librpm function will be used in context part of libdnf (e.g. by microdnf) and in the DNF.

`rpmdbCookie` funkcion is safer. It also detect reinstalation of package as a change. That will be needed in the future to determining if the libsolv cache is still valid. It also solves the libdnf problem with SHA1 in FIPS mode.

Second commit removes class `SHA1Hash`, which is no longer used. And removes OpenSSL require. The class was used by the `dnf_sack_get_rpmdb_version` function, which was removed.

= changelog =
msg: Use `rpmdbCookie` from librpm, remove `hawkey.Sack._rpmdb_version`
type: bugfix
related: https://bugzilla.redhat.com/show_bug.cgi?id=2043476

jrohel committed:
Use `rpmdbCookie` from librpm, remove `hawkey.Sack._rpmdb_version`

`dnf_sack_get_rpmdb_version` function that computed the hash of installed
packages was removed. `rpmdbCookie` function from librpm is used in context
part of libdnf instead.

The private Python function `hawkey.Sack._rpmdb_version()` was removed too.
It was used by DNF. The new version of DNF uses
`rpm.TransactionSet.dbCookie()` function from librpm, which wraps
the `rpmdbCookie` function. So the same librpm function will be used
in context part of libdnf (e.g. by microdnf) and in the DNF.

`rpmdbCookie` funkcion is safer. It also detect reinstalation of package
as a change. That will be needed in the future to determining if the libsolv
cache is still valid.
It also solves the libdnf problem with SHA1 in FIPS mode.

= changelog =
msg: Use `rpmdbCookie` from librpm, remove `hawkey.Sack._rpmdb_version`
type: bugfix
related: https://bugzilla.redhat.com/show_bug.cgi?id=2043476

j-mracek commented on libdnf/dnf-transaction.cpp:
```diff
@@ -1481,14 +1473,8 @@ dnf_transaction_commit(DnfTransaction *transaction, HyGoal goal, DnfState *state
-
-    swdb->endTransaction(_get_current_time(), rpmdb_end.c_str(), libdnf::TransactionState::DONE);
+    rpmdb_end.reset(rpmdbCookie(rpmtsGetRdb(priv->ts)));
+    swdb->endTransaction(_get_current_time(), rpmdb_end.get(), libdnf::TransactionState::DONE);
```
In case that `rpmdb_end` will be `nullptr` it will end on crash in `std::string` constructor. `rpmdbCookie` can return nullptr when  failed.

jrohel replied:
I fixed it.

j-mracek commented on libdnf/dnf-transaction.cpp:
```diff
@@ -1432,16 +1432,8 @@ dnf_transaction_commit(DnfTransaction *transaction, HyGoal goal, DnfState *state
-    }
-    swdb->beginTransaction(_get_current_time(), rpmdb_begin, "", priv->uid);
+    rpmdb_begin.reset(rpmdbCookie(rpmtsGetRdb(priv->ts)));
+    swdb->beginTransaction(_get_current_time(), rpmdb_begin.get(), "", priv->uid);
```
In case that `rpmdb_end` will be `nullptr` it will end on crash in `std::string` constructor. `rpmdbCookie` can return nullptr when  failed.

jrohel replied:
I fixed it.

j-mracek commented:
Please could you also modify `%global dnf_conflict 4.3.0` in `libdnf.spec` to `%global dnf_conflict 4.11.0`

jrohel committed:
Remove class `SHA1Hash`, which is no longer used, remove OpenSSL require

The class was used by the `dnf_sack_get_rpmdb_version` function, which was
removed. The `rpmdbCookie` function from librpm is used instead.

jrohel commented:
I increased version and added conflict.
I also added commit, which removes SHA1Hash class and OpenSSL requirement.

j-mracek commented on libdnf/dnf-transaction.cpp:
```diff
@@ -1432,16 +1433,8 @@ dnf_transaction_commit(DnfTransaction *transaction, HyGoal goal, DnfState *state
-        g_object_unref(rpmdb_version_sack);
-    }
-    swdb->beginTransaction(_get_current_time(), rpmdb_begin, "", priv->uid);
+    rpmdb_begin.reset(rpmdbCookie(rpmtsGetRdb(priv->ts)));
```
Can we be sure that the DB is opened? What about to also open it?

jrohel replied:
Good idea. I added a test if the rpm database is open. And opening, if not.

j-mracek commented on libdnf/dnf-transaction.cpp:
```diff
@@ -1431,17 +1432,20 @@ dnf_transaction_commit(DnfTransaction *transaction, HyGoal goal, DnfState *state
-    swdb->beginTransaction(_get_current_time(), rpmdb_begin, "", priv->uid);
+
+    rpmdb_begin.reset(rpmdbCookie(rpmtsGetRdb(priv->ts)));
+    // FIXME get commandline
```
Please can you also report when `rpmdbCookie(rpmtsGetRdb(priv->ts))` returns `NULL`?

jrohel replied:
OK.

j-mracek commented on libdnf/dnf-transaction.cpp:
```diff
@@ -1481,14 +1485,8 @@ dnf_transaction_commit(DnfTransaction *transaction, HyGoal goal, DnfState *state
-    g_object_unref(rpmdb_version_sack);
-
-    swdb->endTransaction(_get_current_time(), rpmdb_end.c_str(), libdnf::TransactionState::DONE);
+    rpmdb_end.reset(rpmdbCookie(rpmtsGetRdb(priv->ts)));
```
Please can you also report when `rpmdbCookie(rpmtsGetRdb(priv->ts))` returns `NULL`?

jrohel replied:
OK.

j-mracek reviewed (approved):
LGTM
